contrib/registry/file/file_watcher.go

```diff
@@ -17,19 +18,26 @@ type Watcher struct {
 	prefix    string            // Watched prefix key, not file name prefix.
 	discovery gsvc.Discovery    // Service discovery.
 	ch        chan gsvc.Service // Changes that caused by inotify.
+	closed    atomic.Value      // Whether the channel has been closed
 }
 
 // Proceed proceeds watch in blocking way.
 // It returns all complete services that watched by `key` if any change.
 func (w *Watcher) Proceed() (services []gsvc.Service, err error) {
-	<-w.ch
-	return w.discovery.Search(context.Background(), gsvc.SearchInput{
-		Prefix: w.prefix,
-	})
+	if w.closed.Load() == nil {
+		<-w.ch
+		return w.discovery.Search(context.Background(), gsvc.SearchInput{
+			Prefix: w.prefix,
+		})
+	}
+	return nil, context.Canceled
 }
 
 // Close closes the watcher.
 func (w *Watcher) Close() error {
-	close(w.ch)
+	if w.closed.Load() == nil {
```

Comment:
It is appropriate using `*gtype.Bool` for attribute `w.closed`, which can use `Cas` to guarantee atomic operation.

Reply:
Oh, I was not very familiar with the core of the framework before this PR. I will modify it later.